transcom/mymove
# B 17854 prime api recalculate authorized end date

## [B-17854](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A865110)

## Summary

This backlog is the 2nd part of a series to update the Authorized End Date automatically for Origin SIT and Destination SIT service items using the new prime api endpoint [/mto-shipments/{mtoShipmentID}/sit-delivery](http://primelocal:3000/swagger-ui/prime.html#/mtoShipment/updateSITDeliveryRequest).

For Origin SIT only this will also calculate the Required Delivery Date.

Business rules for how dates are calculated:

Origin SIT Authorized End Date (referred to as Allowance End Date in the code) should be 5 days after SIT Customer Contact Date or the SIT Departure Date whichever is earlier.

Origin SIT Required Delivery Date (RDD): If SIT Departure Date is before SIT Customer Contact date + 5 days then RDD = SIT Departure Date + GHC Transit Time. 

If SIT Departure Date is on or after SIT Customer Contact Date + 5 days then RDD = SIT Customer Contact Date + GHC Transit Time + 5 days.
**NOTE: If RDD falls on a weekend or a holiday then RDD is adjusted to be the next workday.
**NOTE: GHC Transit Time is determined based on distance and prime estimated weight and is in the DB table ghc_domestic_transit_time

Desitination SIT Authorized End Date should be 5 days after SIT Customer Requested Delivery date or the SIT Departure Date whichever is earlier.

The calculations require a shipment with a current SIT in Origin or Destination with a populated date for SIT Departure Date. (See test steps below for setting up shipments in the necessary state). 

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test
1. Use the following sql to help setup the needed shipments for testing.

/** Select Current Origin SIT items **/
select * from mto_service_items inner join re_services on mto_service_items.re_service_id = re_services.id 
where re_services.code='DOFSIT' and mto_service_items.sit_entry_date < current_date;

/** replace <SERVICE_ID_HERE> with the id for the service in the above query to set the SIT Departure Date **/
update mto_service_items set sit_departure_date = current_date where mto_service_items.id = 'SERVICE_ID_HERE';

/** Select Current Destination SIT items **/
select * from mto_service_items inner join re_services on mto_service_items.re_service_id = re_services.id 
where re_services.code='DDFSIT' and mto_service_items.sit_entry_date < current_date;

/** Get the move code for your updated service item to lookup in office app and get the etag and id for swagger API call
 replace MOVE_ID_HERE with your move ID from previous queries **/
select mov."locator" from moves mov where mov.id = 'MOVE_ID_HERE';

2. Once you have the shipment ID and the eTag for you shipment open the prime api in swagger. 
http://primelocal:3000/swagger-ui/prime.html#/mtoShipment/updateSITDeliveryRequest

3. Enter a customer contact date and a customer requested delivery date (few days after today). 
4. Execute the call. You should receive a 200 success message. 
5. Open the move task order for your move in the office app and the Authorized End Date and Required Delivery Date (for Origin SIT items only) should be updated.

### Backend


## Screenshots
